### PR TITLE
Reload schema definition on SIGHUP

### LINKED ISSUE #475: Restart required for detecting table relations on schema change
I have an `event_group` table with fields `group_id` and `event_id` referencing `events.id` and `groups.id`respectively.

The weird thing is that even if hydrating the event IDs  works perfectly: 

```
GET /event_group?select=events{*}
[
  {
    "events": {
      "doc": {
     ...
       }
  }
]
```

When I try to do the same with `groups` , I receive this error:
```
GET  /event_group?select=groups{*}
 {
  "hint": null,
  "details": null,
  "code": "42703",
  "message": "column groups.group_id does not exist"
}
```
why does it expect groups.group_id and not groups.id? I didn't need a event.event_id for events!
 
Here's the SQL:

```sql

CREATE TABLE "1".event_group
(
  group_id character varying(32) NOT NULL,
  event_id character varying(36),
  CONSTRAINT fk_eid FOREIGN KEY (event_id) REFERENCES "1".events (id) MATCH SIMPLE ON UPDATE CASCADE ON DELETE NO ACTION,
  CONSTRAINT fk_groups FOREIGN KEY (group_id) REFERENCES "1".groups (id) MATCH SIMPLE ON UPDATE NO ACTION ON DELETE NO ACTION,
  CONSTRAINT uq_gid_eid UNIQUE (event_id, group_id)
)


CREATE TABLE "1".events
(
  doc jsonb,
  id character varying(36) NOT NULL DEFAULT uuid_generate_v4(),
  CONSTRAINT pkey_eid PRIMARY KEY (id)
)

CREATE TABLE "1".groups
(
  id character varying(32) NOT NULL DEFAULT uuid_generate_v4(),
  name character varying(128) NOT NULL,
  description character varying(800),
  CONSTRAINT group_pkey PRIMARY KEY (id)
)
```

### PR DESCRIPTION
**Pull request by begriffs**
Fixes #475

I manually verified this works, but it's hard to think how to add an automated test.

### CODE REVIEW

File: src/PostgREST/Main.hs [hunk 1]
```diff
@@ -58,21 +60,28 @@ main = do
...
   pool <- P.acquire (configPool conf, 10, pgSettings)
 
+  result <- P.use pool $ do
+    supported <- isServerVersionSupported
+    unless supported $ error (
+      "Cannot run in this PostgreSQL version, PostgREST needs at least "
+      <> show minimumPgVersion)
+    getDbStructure (cs $ configSchema conf)
+
+  dbStructure <- newIORef $ either (error.show) id result
```
Comment (ruslantalpa) [inline src/PostgREST/Main.hs]: maybe call it refDbStructure like in App.hs

**Comment by ruslantalpa** [comment]
I know there is no big chance for "race conditions" here but what are the advantages of IORef over MVar? My understanding is that MVar is a IORef with "locks attached"

**Comment by begriffs** [comment]
That's a good question about MVar vs IORef. The code on this branch initially used MVar, but it seemed like it wasn't adding any safety for the situation over IORef so I thought I'd use the simplest thing that works. The only perceived advantage is simplicity.

**Comment by ruslantalpa** [comment]
The question is for my own education, not that there is anything wrong with the code.
Simplicity how? In terms of API, mvar and ioref i think are similar. I am guessing you are referring to  the underlying complexity of the ioref vs mvar libs?

**Comment by ruslantalpa** [comment]
Maybe another way to put it is "is there any time a good reason to use ioref vs mvar" and it seems there might be, some discussion about performance here (didn't look too closely) https://www.reddit.com/r/haskell/comments/39ef3y/ioref_vs_mvar_vs_tvar_vs_tmvar/

**Comment by begriffs** [comment]
IORef does less internally and is a little faster, not that it matters much. I found a good comparison between the two abstractions: [1](http://www.kuznero.com/posts/haskell/2015-02-15-common-concurrency-abstractions-IORef-part-1.html), [2](http://www.kuznero.com/posts/haskell/2015-07-31-common-concurrency-abstractions-MVar-part-2.html).

I think we should be safe from concurrency problems because I'm assuming db changes and sighup happen seldom whereas web requests happen often.
* If a sighup happens a split second before a web request it's no big deal if that request uses old state
* If many sighups happen very quickly it doesn't matter which one wins since the db probably hasn't changed structure during that time

Also MVar models the distinction of being full or empty, whereas our IORef obtains a value upon creation and it doesn't make sense for it to be "emptied."

**Comment by diogob** [comment]
I loved the pull request and the thread here, very informative :D

**Pull request merged**